edb/testbase/server.py

```diff
@@ -1524,63 +1529,140 @@ def start_edgedb_server(
     )
 
 
-def get_cases_by_shard(cases, current_shard, total_shards, verbosity, stats):
+def get_cases_by_shard(cases, selected_shard, total_shards, verbosity, stats):
     if total_shards <= 1:
         return cases
 
-    current_shard -= 1
-    new_test_est = 0.1
-    tests_with_est = []
-    new_tests = []
+    selected_shard -= 1  # starting from 0
+    new_test_est = 0.1  # default estimate if test is not found in stats
+    new_setup_est = 1  # default estimate if setup is not found in stats
+
+    # For logging
     total_tests = 0
     selected_tests = 0
-    for tests in cases.values():
-        for test in tests:
-            total_tests += 1
-            est = stats.get(str(test))
-            if est:
-                heapq.heappush(tests_with_est, (-est[0], test))
+    total_est = 0
+    selected_est = 0
+
+    # Priority queue of tests grouped by setup script ordered by estimated
+    # running time of the groups. Order of tests within cases is preserved.
+    tests_by_setup = []
+
+    # Priority queue of individual tests ordered by estimated running time.
+    tests_with_est = []
+
+    # Prepare the source heaps
+    setup_count = 0
+    for case, tests in cases.items():
+        setup_script = getattr(case, 'get_setup_script', lambda: None)()
+        if setup_script and tests:
+            tests_per_setup = []
+            est_per_setup = setup_est = stats.get(
+                'setup::' + case.get_database_name(), (new_setup_est, 0),
+            )[0]
+            for test in tests:
+                total_tests += 1
+                est = stats.get(str(test), (new_test_est, 0))[0]
+                est_per_setup += est
+                tests_per_setup.append((est, test))
+            heapq.heappush(
+                tests_by_setup,
+                (-est_per_setup, setup_count, setup_est, tests_per_setup),
+            )
+            setup_count += 1
+            total_est += est_per_setup
+        else:
+            for test in tests:
+                total_tests += 1
+                est = stats.get(str(test), (new_test_est, 0))[0]
+                total_est += est
+                heapq.heappush(tests_with_est, (-est, total_tests, test))
+
+    target_est = total_est / total_shards  # target running time of one shard
+    shards_est = [(0, shard, set()) for shard in range(total_shards)]
+    cases = collections.OrderedDict()  # output
```

Comment:
nit: unless you're using some specific OrderedDict apis, you can use the plain python dict here (which are ordered too).

Reply:
Oh, interesting! I was only to keep a consistent returning value as:

```python
def get_test_cases(tests):
    result = collections.OrderedDict()
```